static/js/narrow.js

```diff
@@ -198,6 +200,22 @@ export function activate(raw_operators, opts) {
     const start_time = new Date();
 
     reset_ui_state();
+
+    // Since narrow.activate is called directly from various
+    // places in our code without passing through hashchange,
+    // we need to check if the narrow is allowed for spectator here too.
+
+    if (
+        page_params.is_spectator &&
+        raw_operators.length &&
+        raw_operators.some(
+            (raw_operator) => !hash_util.allowed_web_public_narrows.includes(raw_operator.operator),
```

Comment:
Code structure wise, it'd probably be better for this to call a function from the `spectators` module rather than having the operator-filtering logic here.  But it's pretty easy to do a sweep for such refactors we might want to do later.